### PR TITLE
Refactor Rewards request signing helpers

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/35079

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_rewards/core/legacy/media/youtube.cc
Comment (zenparsing) [inline components/brave_rewards/core/legacy/media/youtube.cc:30]: Moved from `bat_helper.cc` so that that file could be removed without affecting this one.

**Review by emerick — APPROVED** [review]

File: components/brave_rewards/core/common/request_signer.cc
Comment (kdenhartog) [inline components/brave_rewards/core/common/request_signer.cc:69]: Does request content contain any additional headers beyond digest and request-target or do we not sign over them?

File: components/brave_rewards/core/common/signer.cc
Comment (kdenhartog) [inline components/brave_rewards/core/common/signer.cc:62]: Does the `1` do something specific here? When I'm looking at the NACL docs, I don't see 1 being used and it looks like the publicKey and secretKey values here are not being derived from the Seed here, but rather being randomly generated [according to the docs](https://nacl.cr.yp.to/sign.html).

File: components/brave_rewards/core/common/signer.cc
Comment (kdenhartog) [inline components/brave_rewards/core/common/signer.cc:67]: Do we plan to use this for anything but the single key? This isn't really a seed phrase in the since that it's usable and storable by a person since it's just RandBytes. I'm wondering if we need this value at all or if it makes sense to just generate the private key and store that rather than needing to keep two secrets.

File: components/brave_rewards/core/common/signer.cc
Comment (kdenhartog) [inline components/brave_rewards/core/common/signer.cc:32]: When do we expect the private keys and seed to be cleaned up in memory?

File: components/brave_rewards/core/endpoint/promotion/post_suggestions_claim/post_suggestions_claim.cc
Comment (kdenhartog) [inline components/brave_rewards/core/endpoint/promotion/post_suggestions_claim/post_suggestions_claim.cc:90]: Do we no longer need to sign the headers here?

**Review by kdenhartog — CHANGES_REQUESTED** [review]
few questions, but looks like it should be mergable soon

File: components/brave_rewards/core/common/signer.cc
Comment (zenparsing) [reply components/brave_rewards/core/common/signer.cc:32]: `Signer` and `RequestSigner` are short-lived (all instances are function-local currently). The recovery seed is typically read from the Preferences file where it is stored.

File: components/brave_rewards/core/common/signer.cc
Comment (zenparsing) [reply components/brave_rewards/core/common/signer.cc:62]: I believe that we are using a custom implementation that makes use of a third parameter.

https://github.com/brave-intl/bat-native-tweetnacl

File: components/brave_rewards/core/common/signer.cc
Comment (zenparsing) [reply components/brave_rewards/core/common/signer.cc:67]: I'm not sure. It might make sense to revisit this at a later time; moving away from the "recovery seed" approach would be a significant engineering effort (it is also used by the `brave_ads` code).

File: components/brave_rewards/core/common/signer.cc
Comment (zenparsing) [reply components/brave_rewards/core/common/signer.cc:1]: I'm happy to create a follow-up issue to modify the `CODEOWNERS` file.

File: components/brave_rewards/core/common/request_signer.cc
Comment (zenparsing) [reply components/brave_rewards/core/common/request_signer.cc:88]: I've updated to use `CHECK` instead of `DCHECK`, since supplying an arbitrary (possibly empty) string is a programmer error.

Comment (kdenhartog) [reply components/brave_rewards/core/common/request_signer.cc:88]: Do we expect a crash here if that happens? https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/checks.md

`CHECK` is the better option, but in a lot of cases we likely want to be handling our errors rather than crashing the whole browser for this.

File: components/brave_rewards/core/common/signer.cc
Comment (kdenhartog) [reply components/brave_rewards/core/common/signer.cc:22]: Let me double check and sleuth through some previous PRs. We may want to modify this to be random and then store it in prefs. I'm happy with doing this as a follow up issue if necessary though.

File: components/brave_rewards/core/common/request_signer.cc
Comment (kdenhartog) [reply components/brave_rewards/core/common/request_signer.cc:69]: > I don't think any of our current endpoint implementations supply additional headers that would need to be signed
How did we determine this? It seems odd to me to only sign part of the request. I'd think it would be useful for us to sign the entire set of content and headers here instead of just selecting a subset. However, I'm not certain if we have any additional headers that will be going with this request.

File: components/brave_rewards/core/common/signer.cc
Comment (kdenhartog) [reply components/brave_rewards/core/common/signer.cc:62]: Hmm, we probably shouldn't have archived that repo then. Let me look into this before we merge so we don't accidentally have key generation working unexpectedly getting merged and left in.

File: components/brave_rewards/core/common/signer.cc
Comment (kdenhartog) [reply components/brave_rewards/core/common/signer.cc:67]: Ahh ok, that answers my question then that this is operating as a piece of entropy for multiple places. In that case, not worth changing this.

File: components/brave_rewards/core/common/signer.cc
Comment (kdenhartog) [reply components/brave_rewards/core/common/signer.cc:32]: Ok, let's do this as a follow up issue to improve things here but we should ideally be clearing from memory as soon as we no longer expect it to be used. To do this [using something like](https://github.com/brave/brave-core/blob/9824e1e55d742225af5de719b9a9693bf8595e89/components/brave_wallet/browser/internal/hd_key.h#L24C43-L24C62) `SecureVector` is a good way to do this.

Here's an example if you want to take a look at how we do this elsewhere: https://github.com/brave/brave-core/blob/9824e1e55d742225af5de719b9a9693bf8595e89/components/brave_wallet/browser/internal/hd_key.cc#L52

File: components/brave_rewards/core/endpoint/promotion/post_suggestions_claim/post_suggestions_claim.cc
Comment (kdenhartog) [reply components/brave_rewards/core/endpoint/promotion/post_suggestions_claim/post_suggestions_claim.cc:90]: From what I'm understanding here, this previously signed all the headers before and now we're just signing the digest of the content body. This seems like we're effectively signing only part of the message now with this change rather than signing all of the headers. Am I misunderstanding how the code works here?

File: components/brave_rewards/core/common/request_signer.cc
Comment (kdenhartog) [reply components/brave_rewards/core/common/request_signer.cc:69]: Yeah this appears like it will create a regression in [this call](https://github.com/brave/brave-core/pull/21467/files#r1445541437) where some of the headers that were previously signed now won't be.

File: components/brave_rewards/core/common/request_signer.cc
Comment (kdenhartog) [reply components/brave_rewards/core/common/request_signer.cc:56]: no issues here. This was meant to be a note for myself as I was going through the review that would be deleted before I published.

File: components/brave_rewards/core/common/signer.cc
Comment (zenparsing) [reply components/brave_rewards/core/common/signer.cc:32]: https://github.com/brave/brave-browser/issues/35248

File: components/brave_rewards/core/common/signer.cc
Comment (zenparsing) [reply components/brave_rewards/core/common/signer.cc:1]: https://github.com/brave/brave-browser/issues/35249

File: components/brave_rewards/core/common/request_signer.cc
Comment (kdenhartog) [reply components/brave_rewards/core/common/request_signer.cc:69]: Nevermind, I made a mistake - the expected behavior is the same where we previously signed just a subset of headers and are still doing the same. https://github.com/brave/brave-core/pull/21467/files#diff-5fc0b225f5d25ee7f514478ed03449ca53059e18ec6da1905f17946d34a2dc4dL35-L65

File: components/brave_rewards/core/endpoint/promotion/post_suggestions_claim/post_suggestions_claim.cc
Comment (kdenhartog) [reply components/brave_rewards/core/endpoint/promotion/post_suggestions_claim/post_suggestions_claim.cc:90]: I just realized my mistake when looking at this again. You're correct this still matches.

**Review by kdenhartog — APPROVED** [review]
approving and will create follow up issues for some things we can do to improve the overall security in this part of the code.

File: components/brave_rewards/core/common/request_signer.cc
Comment (kdenhartog) [reply components/brave_rewards/core/common/request_signer.cc:88]: > Do we expect a crash here if that happens?

Answering my own question here - Yes, we expect this since we do a control flow check in the primary place where this function should be called right before to make sure this won't happen: https://github.com/brave/brave-core/pull/21467/files#diff-ba7911096320bdaf5945e60d1ac17d386a8c006ac251b1a6162b0205d0731122R64-R67

**Pull request merged**
